psi4/share/psi4/fsapt/fsapt.py

```diff
@@ -328,40 +390,110 @@ def printFrag(geom, Z, Q, fragkeys, frags, nuclear_ws, orbital_ws, filename):
 def extractOsaptData(filepath):
 
     vals = {}
-    vals['Elst']  = readBlock('%s/Elst.dat'  % filepath, H_to_kcal_)
-    vals['Exch']  = readBlock('%s/Exch.dat'  % filepath, H_to_kcal_)
-    vals['IndAB'] = readBlock('%s/IndAB.dat' % filepath, H_to_kcal_)
-    vals['IndBA'] = readBlock('%s/IndBA.dat' % filepath, H_to_kcal_)
-    vals['Disp']  = readBlock('%s/Disp.dat'  % filepath, H_to_kcal_)
+    vals['Elst']  = np.array(readBlock('%s/Elst.dat'  % filepath, H_to_kcal_))
+    vals['Exch']  = np.array(readBlock('%s/Exch.dat'  % filepath, H_to_kcal_))
+    vals['IndAB'] = np.array(readBlock('%s/IndAB.dat' % filepath, H_to_kcal_))
+    vals['IndBA'] = np.array(readBlock('%s/IndBA.dat' % filepath, H_to_kcal_))
+    # Read exact F-SAPT0 dispersion data
+    try:
+        vals['Disp'] = readBlock('%s/Disp.dat'  % filepath, H_to_kcal_) # Exact F-SAPT0 Dispersion
+    except FileNotFoundError:
+        print('No exact dispersion present.  Copying & zeroing `Elst.dat`->`Disp.dat`, and proceeding.\n')
+        vals['Disp'] = np.zeros_like(np.array(vals['Elst']))
+
+    # Read empirical F-SAPT0-D dispersion data
+    try:
+        vals['EDisp'] = readBlock('%s/Empirical_Disp.dat'  % filepath, H_to_kcal_) # Exact F-SAPT0 Dispersion
+    except (FileNotFoundError, OSError):
+        vals['EDisp'] = np.zeros_like(np.array(vals['Elst']))
+
+    # For total, only include exact terms
     vals['Total'] = [[0.0 for x in vals['Elst'][0]] for x2 in vals['Elst']]
     for key in ['Elst', 'Exch', 'IndAB', 'IndBA', 'Disp']:
         for k in range(len(vals['Total'])):
             for l in range(len(vals['Total'][0])):
                 vals['Total'][k][l] += vals[key][k][l]
 
+    # Original fsapt code
+    #vals['Disp']  = readBlock('%s/Disp.dat'  % filepath, H_to_kcal_)
+    #vals['Total'] = [[0.0 for x in vals['Elst'][0]] for x2 in vals['Elst']]
+    #for key in ['Elst', 'Exch', 'IndAB', 'IndBA', 'Disp']:
+    #    for k in range(len(vals['Total'])):
+    #        for l in range(len(vals['Total'][0])):
+    #            vals['Total'][k][l] += vals[key][k][l]
+
     return vals
 
-def extractOrder2Fsapt(osapt, wsA, wsB):
+def fragmentD3Disp(d3disp: np.ndarray, frags: Dict[str, Dict[str, List[str]]]) -> Tuple[float, Dict[str, Dict[str, float]]]:
```

Comment:
Same here `fragment_d3_dispersion`, we have a nice opportunity here while we are changing the functions regardless to be more verbose.